---
title: "EDY usage"
author: "Gema Rojas"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{EDY usage}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Purpose of this package

The aim of this package is to help the user to detect individuals who present extreme downregulation of chromosome Y (EDY) from microarray experiments. The EDY is computed by .....

# Get the necessary information to assess EDY

EDY can be computed using `getEDY()` function. The main input is an object of class `ExpressionSet` The dataset passed through this function requires some information including:

- Whether the gene expression (accessed with `exprs`) is in logarithmic scale or not. 
- Which column from the annotation (accessed with `fData`) contains the `Gene symbol`.
- Which column from the metadata (accessed with `pData`) contains the information about the `gender` and the symbol that codes for `males`.
- In case of having a case/control study, which column from the metadata (accessed with `pData`) contains the information about which individuals belongs to control group, and the symbol that codes for control. 

Let us illustrate how to compute EDY for a real data set available at GEO about .... (GEO accesion number GSE73578). Data can be downloaded into R by

```{r data, eval = FALSE}
library(GEOquery)
gse73578 <- getGEO("GSE73578")[[1]]
```

In order to facilitate reproducibility, we have also include this `ExpressionSet` into our package that is loaded by default. We can visualize the content by:

```{r load_eSet}
gse73578
```



```{r info}
# assayData. In this case, the expression is logarithmic
head(exprs(eSet))

# featureData. In this case, the gene symbol is in the column 'Gene Symbol'
head(fData(eSet))

# phenoData. In this example: gender column = 'Sex:ch1', males = 'M'
# We will not consider any control group
head(pData(eSet))
```

## Assess EDY

Now, we can apply `getEDY()` to know which individuals have the condition:

```{r getEDY}
library(EDY)

edy.gse73578 <- getEDY(x = gse73578, gender.var = 'Sex:ch1',
                       male.key ='M', 
                       gene.key = 'Gene Symbol', log = TRUE)

# How many individuals do and do not have EDY:
table(edy.gse73578$EDY)

# Threshold from which below it is considered EDY:
threshold <- edy.gse73578$threshold
threshold

## Plot the expression of chrY realtive to the rest of the genome in each individual:
plot(edy.gse73578)
```

## Missing information

### Gender of individuals

In case your dataset does not contain information about the sex of the individuals, you can assess it using *massiR*.

### Gene Symbol

If your dataset does not contain information about the gene symbol, this package provide a function to get the hgnc (HUGO gene nomenclature committee) symbols from genBank accession numbers. To use it, we need to know which column from `featureData` has the genBank accession numbers.

```{r GB_to_hgnc}
# Example dataset 
data <- getGEO('40732')
eSet <- data[[1]]

# genBank accession numbers 
head(fData(eSet)) # "GB_ACC" 

# Get hgnc symbols
fData(eSet) <- GB_to_hgnc(x = fData(eSet), genBank.col = "GB_ACC")

head(fData(eSet))
```

Now we can use `getEDY()` with `gene.key = "hgnc_symbol"`.

If your dataset does not have gene symbols nor genBank accession numbers, you can try to get this information from other columns in `featureData` using *biomaRt*.

[massiR]: https://bioconductor.org/packages/release/bioc/html/massiR.html
[biomaRt]: https://bioconductor.org/packages/release/bioc/html/biomaRt.html
