---
title: "EDY usage"
author: "Gema Rojas"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{EDY usage}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Purpose of this package

The aim of this package is to help the user to detect individuals who present extreme downregulation of chromosome Y (EDY) from microarray experiments. The EDY is assessed by computing the relative expression of genes in chromosome Y with regard the autosomal genes for each individual, and then, the next formulae is applied to stablish a trheshold from which down to consider that the individual has EDY:
$$Threshold = median - 1.2IQR$$

# Get the necessary information to assess EDY

EDY can be computed using `getEDY()` function. The main input is an object of class `ExpressionSet`. This function requires also some information including:

- Whether the gene expression (accessed with `exprs`) is in logarithmic scale or not. 
- Which column from the annotation (accessed with `fData`) contains the `Gene symbol`.
- Which column from the metadata (accessed with `pData`) contains the information about the `gender` and the symbol that codes for `males`.
- In case of having a case/control study, which column from the metadata (accessed with `pData`) contains the information about which individuals belongs to control group, and the symbol that codes for control. 

Let us illustrate how to compute EDY for a real data set available at GEO. It contains information about gene expression in children with acute lymphoblastic leukemia that has been treated or not with glucocorticoids (GEO accesion number GSE73578). Data can be downloaded into R by

```{r data, eval = FALSE}
library(GEOquery)
gse73578 <- getGEO("GSE73578")[[1]]
```

In order to facilitate reproducibility, we have also include this `ExpressionSet` into our package that is loaded by default. We can visualize the content by:

```{r load_eSet}
gse73578
```


The information about whether the expression data is in logarithmic scale or not should be specified in the description of the data set, but it can also be checked with `exprs`. In this case, it is logarithhmic: 

```{r info}
head(exprs(gse73578))
```

The information about gene symbol, gender and males key must be found with `pData`. In this case, gene symbol column is named `Gene Symbol`, gender column is `Sex:ch1` and males key is `M`:

```{r info2}
head(fData(gse73578))
```

## Assess EDY

Now, we can apply `getEDY`:

```{r getEDY}
library(EDY)

edy.gse73578 <- getEDY(x = gse73578, gender.var = 'Sex:ch1',
                       male.key = 'M', 
                       gene.key = 'Gene Symbol', log = TRUE)

edy.gse73578
```

A scatter plot can be easily generated by `plot` to visualize the relative expression of chromosome Y with regard to autosomal genes for each individual, as well as the threshold (dashed line) and the individuals considered as EDY:

```{r plot}
plot(edy.gse73578)
```

To obtain the identifier of individuals classified as EDY, the following code can be used:

```{r EDY selection}
EDYindividuals <- names(edy.gse73578$EDY[edy.gse73578$EDY=="Yes"])
EDYindividuals
```


# Missing information

### Gender of individuals

In case your dataset does not contain information about the sex of the individuals, you can assess it using *massiR*.

### Gene Symbol

If your dataset does not contain information about the gene symbol, the function `get_hgnc` can be used to get the hgnc (HUGO gene nomenclature committee) symbols from genBank accession numbers, entrezgenes (NCBI gene IDs) or Ensembl stable IDs. 

The required inputs are:

- An ExpressionSet.
- The type of gene ID that we want to use to obtain the HGNC symbols. Must be one of **genebank**, **entrezgene** or **ensembl**.
- The name of the column that contains the gene IDs. It can be accessed by `fData`.

In the example data set, we can use for instance entregene IDs:

```{r get_hgnc entrezgene}
head(fData(gse73578))  

# Get HGNC symbols from entrezgene IDs:
fData(gse73578) <- get_hgnc(x = gse73578, gene.id = "entrezgene", gene.col = "ENTREZ_GENE_ID")
head(fData(gse73578))
```

Similarly, in this example we could have used genBak accession numbers:

```{r genbank, eval=FALSE}
# Get HGNC symbols from genBank accession numbers:
fData(gse73578) <- GB_to_hgnc(x = gse73578, gene.id = "genbank", gene.col = "GB_ACC")
```

Now `getEDY` can be used with `gene.key = "hgnc_symbol"`:

```{r getEDY hgnc symbol}
edy.gse73578_2 <- getEDY(x = gse73578, gender.var = 'Sex:ch1',
                       male.key = 'M', 
                       gene.key = 'hgnc_symbol', log = TRUE)

edy.gse73578_2
plot(edy.gse73578_2)
```

If your dataset does not have gene symbols, genBank accession numbers, entrezgene IDs nor Ensembl stable IDs, you can try to get this information from other columns in `featureData` using *biomaRt*.

[massiR]: https://bioconductor.org/packages/release/bioc/html/massiR.html
[biomaRt]: https://bioconductor.org/packages/release/bioc/html/biomaRt.html
