---
title: "EDY usage"
author: "Gema Rojas"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{EDY usage}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Purpose of this package

The aim of this package is to help the user to detect individuals who present extreme downregulation of chromosome Y (EDY) in microarray experiments.

## Get the necessary information to assess EDY

In order to know which individuals have EDY, we need to look for some information in our data. For this vignette, we will use the following dataset as an example:

```{r data, echo = FALSE}
# Loading data (microarray ExpresionSet)
library('GEOquery')
data <- getGEO('GSE73578')
eSet <- data[[1]]
```

We need to know:

- Wether the gene expression in `assayData` is given in a logarithmic scale or not. 
- Which column from `featureData` contains the `Gene symbol`.
- Which column from `phenoData`contains the information about the `gender` and the symbol that codes for `males`.
- Which column from `phenoData` contains the information about which individuals belongs to control group, and the symbol that codes for control. 

```{r info}
# assayData. In this case, the expression is logarithmic
head(exprs(eSet))

# featureData. In this case, the gene symbol is in the column 'Gene Symbol'
head(fData(eSet))

# phenoData. In this example: gender column = 'Sex:ch1', males = 'M'
# We will not consider any control group
head(pData(eSet))
```

## Assess EDY

Now, we can apply `getEDY(...)` to know which individuals have the condition:

```{r getEDY}
library('EDY')

output.GSE73578 <- getEDY(x = eSet, gender.var = 'Sex:ch1',
                       male.key ='M', 
                       gene.key = 'Gene Symbol', log = T)

# How many individuals do and do not have EDY:
table(output.GSE73578$EDY)

# Threshold from which below it is considered EDY:
threshold <- out.GSE73578$threshold
threshold

## Plot the expression of chrY realtive to the rest of the genome in each individual:
plot(apply(out.GSE73578$Ry, 1, mean), xlab = "Individuals", ylab = "chrY / autosomomes")
abline(h=threshold, col="blue")
```

## Missing information

### Gender of individuals

In case your dataset does not contain information about the sex of the individuals, you can assess it using *massiR*.

### Gene Symbol

If your dataset does not contain information about the gene symbol, this package provide a function to get the hgnc (HUGO gene nomenclature committee) symbols from genBank accession numbers. To use it, we need to know which column from `featureData` has the genBank accession numbers.

```{r GB_to_hgnc}
# Example dataset 
data2 <- getGEO('40732')
eSet2 <- data2[[1]]

# genBank accession numbers 
head(fData(eSet2)) # "GB_ACC" 

# Get hgnc symbols
fData(eSet2) <- GB_to_hgnc(x = fData(eSet2), genBank.col = "GB_ACC")

head(fData(eSet2))
```

Now we can use `getEDY()` with `gene.key = "hgnc_symbol"`.

If your dataset does not have gene symbols nor genBank accession numbers, you can try to get this information from other columns in `featureData` using *biomaRt*.

[massiR]: https://bioconductor.org/packages/release/bioc/html/massiR.html
[biomaRt]: https://bioconductor.org/packages/release/bioc/html/biomaRt.html
