---
title: "Extreme donwregulation of chromosome Y"
author: 
- name: Gema Rojas
  affiliation: 
  - &uab Autonomous University of Barcelona (UAB)
  - &isglobal ISGlobal, Centre for Research in Environmental Epidemiology (CREAL)
  - &brge Bioinformatics Research Group in Epidemiolgy (BRGE)
- name: Alejandro CÃ¡ceres
  affiliation: 
  - *isglobal
  - *brge
- name: Juan R. Gonzalez
  affiliation: 
  - *isglobal
  - *brge
  - *uab
  email: juanr.gonzalez@isglobal.org
date: "`r Sys.Date()`"
package: "`r pkg_ver('EDY')`"
output: 
  BiocStyle::html_document:
    number_sections: true
    toc: yes
    fig_caption: yes
    toc_float: true
vignette: >
  %\VignetteIndexEntry{Extreme donwregulation of chromosome Y}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r setup, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "", 
                      cache=TRUE, message = FALSE
)
```

# Purpose of this package

The aim of this package is to help the user to detect individuals who present extreme downregulation of chromosome Y (EDY) from microarray or RNA-seq experiments. For each individual, we measured the relative expression of the entire chromosome Y with respect to the autosomes. For $n$ probes (exons) in
chromosome Y, with $x_i$ intensity (read counts) for the $i$-th probe, we compute 
$y=1/n \sum_{i=1}^N \log_2(x_i)$ as a measure of the average expression of chromosome Y. Likewise, for $m$ probes in the autosomes, we computed the mean expression of autosomes $a = 1/m \sum_{i=1}^M \log_2(x_i)$ [NOTE: for RNAseq data $\log_2(x_i + 1)$ is computed to avoid problems with zero counts]. The
relative amount of an individual's Y expression with respect to the individual's autosomes was then
$Ry=y-a$, and, in a population sample, the individual $j$ was considered with EDY if

$$Ry_j < median(Ry) - 1.2 IQR(Ry)$$, 

where IQR is the inter-quartile range across the sample.


# Get the necessary information to assess EDY

First, let us load the required packages

```{r load_packages}
library(EDY)
library(caret)
library(tidyverse)
library(Biobase)
```

EDY can be computed using `getEDY()` function from `EDY` package. The main input is an object of class `ExpressionSet`. This function requires also some information, including:

- Whether the gene expression (accessed with `exprs`) is in logarithmic scale or not. 
- Which column from the annotation (accessed with `fData`) contains the `Gene symbol`. This is required to get the relative expression of chromosome Y.
- Which column from the metadata (accessed with `pData`) contains the information about the `gender` and the symbol that codes for `males`. This is required since EDY is computed only in male samples.
- In case of having a case/control study, which column from the metadata (accessed with `pData`) contains the information about the groups, and the symbol that codes for control. When this information is available, the $Ry$ value is estimated only in controls. 

Let us illustrate how to compute EDY for a real data set available at GEO. It contains information about gene expression in children with acute lymphoblastic leukemia that has been treated or not with glucocorticoids (GEO accesion number GSE73578). Data can be downloaded into R by:

```{r data, eval = FALSE}
library(GEOquery)
gse73578 <- getGEO("GSE73578")[[1]]
```

In order to facilitate reproducibility, we have also include this `ExpressionSet` into our package and it is loaded by default. We can visualize the content by:

```{r load_eSet}
gse73578
```

The information about whether the expression data is in logarithmic scale or not should be specified in the description of the data set, but it can also be checked with `exprs`. In this case, it is logarithmic: 

```{r info}
exprs(gse73578)[1:5, 1:5]
```

The information about gene symbol can be accessed by `fData`. If you cannot find it there, please see [Missing information](#missing-information) section below. In this case, the gene symbol column is named `Gene Symbol`:

```{r info2}
fData(gse73578)[1:5,1:4]
names(fData(gse73578))
```


In some occasions, the same entry has several gene symbols. In this example, the gene symbols are separted by ` /// `. To select only the first symbol we can use the following code, otherwise, `getEDY` will not recognise the transcript:

```{r split_symbols}
tmp <- strsplit(fData(gse73578)[, "Gene Symbol"], " /// ")
fData(gse73578)$Symbol <- unlist(lapply(tmp, function(x) x[1]))

head(fData(gse73578)[,c("Gene Symbol","Symbol")])
```

The information about the sex of individuals must be found using `pData`. In this case, gender column is called `Sex:ch1` and males key is `M`. If, for some reason, your data do not contain this information, this can be estimated from expression data (see [Missing information](#missing-information) section below).

```{r info3}
names(pData(gse73578))
```

This is the total number of male and female samples

```{r tableEDY0}
table(gse73578$`Sex:ch1`)
```


## Assess EDY

Once this information is obtained, the EDY status of **male** samples can be obtained by:

```{r getEDY}
edy.gse73578 <- getEDY(x = gse73578, gender.var = 'Sex:ch1',
                       male.key = 'M', 
                       gene.key = 'Symbol', experiment.type = "microarray")

edy.gse73578
```

Notice that the argument `log=TRUE` has be omitted in the function since this is the default value. This function returns an `ExpressionSet` object having EDY status in a new variable called `EDY` 

```{r tableEDY}
table(edy.gse73578$EDY)
```

A scatter plot can be easily generated by using the generic function `plot`. It allows to visualize the relative expression of chromosome Y with regard to autosomal genes for each individual, as well as the threshold (dashed line) and the individuals considered as EDY:

```{r plot}
plot(edy.gse73578)
```

The ids of individuals classified as EDY can be obtained by 

```{r EDY_selection}
EDYindividuals <- names(edy.gse73578$EDY[edy.gse73578$EDY=="Yes"])
EDYindividuals
```


# Missing information

In this section we are describing how to obtain some of the required information to estimate EDY. It includes examples about how to get gender of individuals or gene symbol that is required to select probes from chromosome Y.

### Gender of individuals

In case your dataset does not contain information about the sex of the individuals, you can assess it using [massiR].

### Gene Symbol

If your dataset does not contain information about the gene symbol, the function `get_hgnc` can be used to get the hgnc (HUGO gene nomenclature committee) symbols from genBank accession numbers, entrezgenes (NCBI gene IDs) or other types of identifiers. This function uses the package [org.Hs.eg.db] when we want to obtain the symbols from genBank accession numbers. For the rest of the cases, it uses the package [EnsDb.Hsapiens.v86]. [Here] you can find more information about how to use the package and explore it contents.


The required inputs are:

- An ExpressionSet.
- The type of gene ID that we want to use to obtain the HGNC symbols. Must be one of It must be one of `ENTREZID`, `EXONID`, `GENEBIOTYPE`, `GENEID`, `GENENAME`, `PROTDOMID`, `PROTEINDOMAINID`, `PROTEINDOMAINSOURCE`, `PROTEINID`, `SEQNAME`, `SEQSTRAND`, `TXBIOTYPE`, `TXID`, `TXNAME`, `UNIPROTID` or `GENBANK`.
- The name of the column that contains the gene IDs. It can be accessed by `fData`.

In the example data set, we can use for instance entrezgene IDs. If  the same entry has several entrezgene IDs, we need to select just one like we did before:

```{r entrezgene}
head(fData(gse73578)[,"ENTREZ_GENE_ID"])  

tmp <- strsplit(fData(gse73578)[, "ENTREZ_GENE_ID"], " /// ")
fData(gse73578)$ENTREZ_GENE_ID <- unlist(lapply(tmp, function(x) x[1]))

head(fData(gse73578)[,"ENTREZ_GENE_ID"]) 
```

```{r get_hgnc_entrezgene}
# Get HGNC symbols from entrezgene IDs:
gse73578 <- get_hgnc(x = gse73578, key.type = "entrezid", key.col = "ENTREZ_GENE_ID")
fData(gse73578)[1:5, 1:5]
```

Observe that column `hgnc_symbol` has been added to the annotation. Similarly, in this example we could have used genBank accession numbers by simply:

```{r genbank, eval=FALSE}
# Get HGNC symbols from genBank accession numbers:
gse73578 <- get_hgnc(x = gse73578, key.type = "genbank", key.col = "GB_ACC")
```

Now `getEDY` can be applied with `gene.key = "hgnc_symbol"`:

```{r getEDY_hgnc_symbol}
edy.gse73578_2 <- getEDY(x = gse73578, gender.var = 'Sex:ch1',
                       male.key = 'M', 
                       gene.key = 'hgnc_symbol', experiment.type = "microarray")

edy.gse73578_2
plot(edy.gse73578_2)
```

```{r tableEDY2}
table(edy.gse73578_2$EDY)
```


```{r EDY selection2}
EDYindividuals2 <- names(edy.gse73578_2$EDY[edy.gse73578_2$EDY=="Yes"])
EDYindividuals2
```

As we can see, the results are not the same as when using gene symbols from the original data set. It is due to the loss of some gene IDs during the translation, since not all the IDs find their equivalence in the consulted databases.

If your dataset does not have gene symbols, genBank accession numbers, entrezgene IDs nor any of the accepted identifiers, you can try to get this information from other columns in `fData` using [biomaRt].

# EDY prediction from methylation data.

Another utility of this package it to predict extreme downregulation of chromosome Y using methylation data. For this purpose, we use the function `predictEDY`. The only input for this function is *one* of: 

- A matrix with CpGs in rows, samples in columns and CpG names in the rowname.
- A data.frame with individuals/samples in columns, being the first column the CpG names.
- An ExpressionSet with methylation data.

The input must contain only male individuals, since it does not make sense to do a prediction of EDY in females. If it contains NAs, some predictions may be NA too. 

The prediction is based on an Elastic Net regression model, trained with data from [The Cancer Genome Atlas (TCGA)]. In the following example, we use a matrix containing methylation information of chromosome Y of men who suffer from  localized, non-indolent prostate cancer (subset from `GSE84043`). The matrix is available in this package and loaded by default to facilitate reproducibility:

```{r predictEDY}
head(chrY.meth.matrix)[,1:6]
edy.prediction <- predictEDY(chrY.meth.matrix)
```

The accuracy and the confidence interval 95% of the model is calculated by using the same CpGs of chromosome Y in your data to predict EDY in the TCGA dataset and then comparing with the real EDY calculated with `getEDY`. Next, the model is used to predict EDY in the problem matrix. The results are the following:

```{r prediction results}
edy.prediction[1:20]
table(edy.prediction)
```

[massiR]: https://bioconductor.org/packages/release/bioc/html/massiR.html
[biomaRt]: https://bioconductor.org/packages/release/bioc/html/biomaRt.html
[The Cancer Genome Atlas (TCGA)]: https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga
[EnsDb.Hsapiens.v86]: http://bioconductor.org/packages/release/data/annotation/manuals/EnsDb.Hsapiens.v86/man/EnsDb.Hsapiens.v86.pdf
[org.Hs.eg.db]: https://bioconductor.org/packages/release/data/annotation/manuals/org.Hs.eg.db/man/org.Hs.eg.db.pdf
[Here]: http://bioconductor.org/packages/release/bioc/vignettes/ensembldb/inst/doc/ensembldb.html#8_using_ensdb_objects_in_the_annotationdbi_framework